sqlglot/transforms.py

```diff
@@ -79,6 +79,35 @@ def eliminate_distinct_on(expression: exp.Expression) -> exp.Expression:
     return expression
 
 
+def eliminate_qualify(expression: exp.Expression) -> exp.Expression:
+    """
+    Convert SELECT statements that contain the QUALIFY clause into subqueries, filtered equivalently.
+
+    The idea behind this transformation can be seen in Snowflake's documentation for QUALIFY:
+    https://docs.snowflake.com/en/sql-reference/constructs/qualify
+
+    Some dialects don't support window functions in the WHERE clause -- we need to add them in the
+    subquery's projection list so they can be referenced in the outer filter using an alias. if the
+    selected columns are known (i.e. no "*"), we can simply copy them to the outer query without
+    including the window function.
+    """
+
+    if isinstance(expression, exp.Select) and expression.args.get("qualify"):
+        outer_selects = exp.select(*[s.alias_or_name or s for s in expression.selects])
+        qualify_filters = expression.args["qualify"].this
+        expression.args["qualify"].pop()
+        sequence = itertools.count()
+
+        for window in qualify_filters.find_all(exp.Window):
+            window_alias = f"_w_{next(sequence)}"
+            expression.select(exp.alias_(window.copy(), window_alias), copy=False)
```

Comment:
does transform auto copy expression? or are you mutating it here

Reply:
`preprocess` copies the expression before applying any transformations.